### ISSUE TITLE
Schema Evolution With Custom Long Types (throws error)

### ISSUE THREAD

**Issue by KeithWoods** [mtth/avsc#143]
First up, fantastic library. It's been super solid and just gets on with it, well done :)

We've got some odd behaviour around deserialising when it involves an evolved schema and the schema in question has a custom long type.

It appears that when using a resolver to deserialize this line blows up: 
https://github.com/mtth/avsc/blob/master/lib/types.js#L2644
At this point `this` is actually a resolver which doesn't have a `_fromBuffer` function (thought it does have `this._readerType._fromBuffer`. Not sure how to work around it at this point. Any help or ideas would be much appriciated.

I've put together 2 near identical examples, both deal with schema evolution, one with custom long types and one without.  

## They share this package.json:
```
{
  "name": "avsc-issue",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "avsc": "^5.1.1",
    "long": "^3.2.0"
  }
}
```

## evolutionWithCustomLongTypes.js
This example blows up as described within/below.

```
var avsc = require('avsc');
var Long = require('long');

// custom long type copied from https://github.com/mtth/avsc/wiki/Advanced-usage#custom-long-types
var longType = avsc.types.LongType.__with({
  fromBuffer: (buf) => {
    return new Long(buf.readInt32LE(), buf.readInt32LE(4));
  },
  toBuffer: (n) => {
    var buf = Buffer.alloc(8);
    buf.writeInt32LE(n.getLowBits());
    buf.writeInt32LE(n.getHighBits(), 4);
    return buf;
  },
  fromJSON: Long.fromValue,
  toJSON: (n) => { return +n; },
  isValid: Long.isLong,
  compare: (n1, n2) => { return n1.compare(n2); }
});

// first version of our type/schema
var decimalTypeV1 = avsc.Type.forSchema(
	{	
		"type": "record",
		"name": "DecimalDto",
		"namespace": "my.amespace",
		"fields": [
			{"name": "unscaledValue", "type": "long"}, 
			{"name": "scale", "type": "int"}
		]
	}, 
	{'registry': {'long': longType} }
);

// second version of our type/schema, adds an optional field
var decimalTypeV2 = avsc.Type.forSchema(
	{	
		"type": "record",
		"name": "DecimalDto",
		"namespace": "my.amespace",
		"fields": [
			{"name": "unscaledValue", "type": "long"}, 
			{"name": "scale", "type": "int"}, 
			{"name":"newField","type":["null",{"type":"string","avro.java.string":"String"}],"default":null}
		]
	}, 
	{'registry': {'long': longType} }
);

var decimalDto = {
	unscaledValue: Long.fromNumber(100000),
	scale: 4,
	newField: 'foo'
};

var buffer = decimalTypeV2.toBuffer(decimalDto);


try {
    console.log('\n** Works, we\'re just deserialising with the same type/schema we serialised with');
    console.log(
        decimalTypeV2.fromBuffer(buffer)
    );
} catch (err) {
    console.log(err);
}

try {
    console.log('\n**  Won\'t work as the schema has moved on, it\'s not a bug, we need both v1 and v2 schemas to deserialise (next example)');
    console.log(
        decimalTypeV1.fromBuffer(buffer)
    );
} catch (err) {
    console.log(err);
}

try {
    console.log('\n**  This is the case where you need evolution. Works with there are no custom long types else blows up');
    let resolver = decimalTypeV1.createResolver(decimalTypeV2);
    console.log(
        decimalTypeV1.fromBuffer(buffer, resolver, true)
    );
} catch (err) {
    console.log(err);
}
```
output: 
```
$ node evolutionWithCustomLongTypes.js

** Works, we're just deserialising with the same type/schema we serialised with
DecimalDto {
  unscaledValue: Long { low: 100000, high: 0, unsigned: false },
  scale: 4,
  newField: 'foo' }

**  Won't work as the schema has moved on, it's not a bug, we need both v1 and v2 schemas to deserialise (next example)
Error: trailing data
    at RecordType.Type.fromBuffer (C:\workspace\dev\avsc-issue\node_modules\avsc\lib\types.js:578:11)
    at Object.<anonymous> (C:\workspace\dev\avsc-issue\evolutionWithCustomLongTypes.js:71:23)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)

**  This is the case where you need evolution. Works with there are no custom long types else blows up
TypeError: this._fromBuffer is not a function
    at Resolver.AbstractLongType._read (C:\workspace\dev\avsc-issue\node_modules\avsc\lib\types.js:2644:17)
    at Resolver.readDecimalDto [as _read] (eval at <anonymous> (C:\workspace\dev\avsc-issue\node_modules\avsc\lib\types.js:2375:20), <anonymous>:4:28)
    at readValue (C:\workspace\dev\avsc-issue\node_modules\avsc\lib\types.js:2786:21)
    at RecordType.Type.fromBuffer (C:\workspace\dev\avsc-issue\node_modules\avsc\lib\types.js:573:13)
    at Object.<anonymous> (C:\workspace\dev\avsc-issue\evolutionWithCustomLongTypes.js:81:23)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
```
## evolutionWithoutCustomLongTypes.js

This works as expected, just providing for full context. It's the same as above with the long type removed and in the schema `long` is replaced with `int`, 

```
let avsc = require('avsc');

// first version of our type/schema
let decimalTypeV1 = avsc.Type.forSchema(
    {
        "type": "record",
        "name": "DecimalDto",
        "namespace": "my.amespace",
        "fields": [
            {"name": "unscaledValue", "type": "int"},
            {"name": "scale", "type": "int"}
        ]
    }
);

// second version of our type/schema, adds an optional field
let decimalTypeV2 = avsc.Type.forSchema(
    {
        "type": "record",
        "name": "DecimalDto",
        "namespace": "my.amespace",
        "fields": [
            {"name": "unscaledValue", "type": "int"},
            {"name": "scale", "type": "int"},
            {"name": "newField", "type": ["null", {"type": "string", "avro.java.string": "String"}], "default": null}
        ]
    }
);

let decimalDto = {
    unscaledValue: 10,
    scale: 4,
    newField: 'foo'
};

let buffer = decimalTypeV2.toBuffer(decimalDto);

try {
    console.log('\n** Works, we\'re just deserialising with the same type/schema we serialised with');
    console.log(
        decimalTypeV2.fromBuffer(buffer)
    );
} catch (err) {
    console.log(err);
}

try {
    console.log('\n**  Won\'t work as the schema has moved on, it\'s not a bug, we need both v1 and v2 schemas to deserialise (next example)');
    console.log(
        decimalTypeV1.fromBuffer(buffer)
    );
} catch (err) {
    console.log(err);
}

try {
    console.log('\n**  This is the case where you need evolution. Works with there are no custom long types');
    let resolver = decimalTypeV1.createResolver(decimalTypeV2);
    console.log(
        decimalTypeV1.fromBuffer(buffer, resolver, true)
    );
} catch (err) {
    console.log(err);
}
```

output:

```
$ node evolutionWithoutCustomLongTypes.js

** Works, we're just deserialising with the same type/schema we serialised with
DecimalDto { unscaledValue: 10, scale: 4, newField: 'foo' }

**  Won't work as the schema has moved on, it's not a bug, we need both v1 and v2 schemas to deserialise (next example)
Error: trailing data
    at RecordType.Type.fromBuffer (C:\workspace\dev\avsc-issue\node_modules\avsc\lib\types.js:578:11)
    at Object.<anonymous> (C:\workspace\dev\avsc-issue\evolutionWithoutCustomLongTypes.js:50:23)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)

**  This is the case where you need evolution. Works with there are no custom long types
DecimalDto { unscaledValue: 10, scale: 4 }

```

Any idea how we can get evolution with custom long types? Thanks again for any help,

**Comment by mtth** [comment]
Hey @KeithWoods, thank you for the detailed explanation and examples! It's a bug in the evolution logic for custom long types; I'm aiming to have the fix out by tonight.

**Comment by mtth** [comment]
This went more smoothly than expected: the fix is now released in `5.1.2`. Please let me know if you encounter any more issues.

**Comment by KeithWoods** [comment]
Great turnaround. Thanks so much.
